File: Go/VerbalArithmeticPuzzle.go
//go:build ignore

package main

import "fmt"

func isSolvable(words []string, result string) bool {
	maxLen := 0
	for _, w := range words {
		if len(w) > maxLen {
			maxLen = len(w)
		}
	}
	if maxLen > len(result) {
		return false
	}

	revWords := make([]string, len(words))
	for i, w := range words {
		b := []byte(w)
		for l, r := 0, len(b)-1; l < r; l, r = l+1, r-1 {
			b[l], b[r] = b[r], b[l]
		}
		revWords[i] = string(b)
	}
	revResult := []byte(result)
	for l, r := 0, len(revResult)-1; l < r; l, r = l+1, r-1 {
		revResult[l], revResult[r] = revResult[r], revResult[l]
	}

	leading := [26]bool{}
	for _, w := range words {
		if len(w) > 1 {
			leading[w[0]-'A'] = true
		}
	}
	if len(result) > 1 {
		leading[result[0]-'A'] = true
	}

	const unset = -1
	var charDigit [26]int
	for i := range charDigit {
		charDigit[i] = unset
	}
	var used [10]bool

	var dfs func(col, idx, carry int) bool
	dfs = func(col, idx, carry int) bool {
		if col == len(revResult) {
			return carry == 0
		}
		if idx < len(revWords) {
			w := revWords[idx]
			if col >= len(w) || charDigit[w[col]-'A'] != unset {
				return dfs(col, idx+1, carry)
			}
			ch := w[col] - 'A'
			for d := range 10 {
				if used[d] || (d == 0 && leading[ch]) {
					continue
				}
				charDigit[ch] = d
				used[d] = true
				if dfs(col, idx+1, carry) {
					return true
				}
				charDigit[ch] = unset
				used[d] = false
			}
			return false
		}

		colSum := carry
		for _, w := range revWords {
			if col < len(w) {
				colSum += charDigit[w[col]-'A']
			}
		}
		digit := colSum % 10
		nextCarry := colSum / 10
		rCh := revResult[col] - 'A'
		if charDigit[rCh] != unset {
			if charDigit[rCh] != digit {
				return false
			}
			return dfs(col+1, 0, nextCarry)
		}
		if used[digit] || (digit == 0 && leading[rCh]) {
			return false
		}
		charDigit[rCh] = digit
		used[digit] = true
		if dfs(col+1, 0, nextCarry) {
			return true
		}
		charDigit[rCh] = unset
		used[digit] = false
		return false
	}

	return dfs(0, 0, 0)
}

func main() {
	fmt.Println(isSolvable([]string{"SEND", "MORE"}, "MONEY"))           // true
	fmt.Println(isSolvable([]string{"SIX", "SEVEN", "SEVEN"}, "TWENTY")) // true
	fmt.Println(isSolvable([]string{"LEET", "CODE"}, "POINT"))           // false
	fmt.Println(isSolvable([]string{"A", "B"}, "C"))                     // true
	fmt.Println(isSolvable([]string{"ACA"}, "JA"))                       // false
}

File: Rust/src/bin/verbal_arithmetic_puzzle.rs
struct Solver {
    rev_words: Vec<Vec<u8>>,
    rev_result: Vec<u8>,
    leading: [bool; 26],
    char_digit: [i32; 26],
    used: [bool; 10],
}

const UNSET: i32 = -1;

impl Solver {
    fn dfs(&mut self, col: usize, idx: usize, carry: i32) -> bool {
        if col == self.rev_result.len() {
            return carry == 0;
        }
        if idx < self.rev_words.len() {
            let w = &self.rev_words[idx];
            if col >= w.len() || self.char_digit[(w[col] - b'A') as usize] != UNSET {
                return self.dfs(col, idx + 1, carry);
            }
            let ch = (w[col] - b'A') as usize;
            for d in 0..10 {
                if self.used[d] || (d == 0 && self.leading[ch]) {
                    continue;
                }
                self.char_digit[ch] = d as i32;
                self.used[d] = true;
                if self.dfs(col, idx + 1, carry) {
                    return true;
                }
                self.char_digit[ch] = UNSET;
                self.used[d] = false;
            }
            return false;
        }

        let mut col_sum = carry;
        for w in &self.rev_words {
            if col < w.len() {
                col_sum += self.char_digit[(w[col] - b'A') as usize];
            }
        }
        let digit = (col_sum % 10) as usize;
        let next_carry = col_sum / 10;
        let r_ch = (self.rev_result[col] - b'A') as usize;
        if self.char_digit[r_ch] != UNSET {
            if self.char_digit[r_ch] as usize != digit {
                return false;
            }
            return self.dfs(col + 1, 0, next_carry);
        }
        if self.used[digit] || (digit == 0 && self.leading[r_ch]) {
            return false;
        }
        self.char_digit[r_ch] = digit as i32;
        self.used[digit] = true;
        if self.dfs(col + 1, 0, next_carry) {
            return true;
        }
        self.char_digit[r_ch] = UNSET;
        self.used[digit] = false;
        false
    }
}

fn is_solvable(words: Vec<String>, result: String) -> bool {
    let max_len = words.iter().map(|w| w.len()).max().unwrap_or(0);
    if max_len > result.len() {
        return false;
    }
    let rev_words: Vec<Vec<u8>> = words
        .iter()
        .map(|w| w.bytes().rev().collect())
        .collect();
    let rev_result: Vec<u8> = result.bytes().rev().collect();

    let mut leading = [false; 26];
    for w in &words {
        if w.len() > 1 {
            leading[(w.as_bytes()[0] - b'A') as usize] = true;
        }
    }
    if result.len() > 1 {
        leading[(result.as_bytes()[0] - b'A') as usize] = true;
    }

    let mut solver = Solver {
        rev_words,
        rev_result,
        leading,
        char_digit: [UNSET; 26],
        used: [false; 10],
    };
    solver.dfs(0, 0, 0)
}

fn s(strs: &[&str]) -> Vec<String> {
    strs.iter().map(|x| x.to_string()).collect()
}

fn main() {
    println!("{}", is_solvable(s(&["SEND", "MORE"]), "MONEY".to_string()));            // true
    println!("{}", is_solvable(s(&["SIX", "SEVEN", "SEVEN"]), "TWENTY".to_string()));  // true
    println!("{}", is_solvable(s(&["LEET", "CODE"]), "POINT".to_string()));            // false
    println!("{}", is_solvable(s(&["A", "B"]), "C".to_string()));                      // true
    println!("{}", is_solvable(s(&["ACA"]), "JA".to_string()));                        // false
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn examples() {
        assert!(is_solvable(s(&["SEND", "MORE"]), "MONEY".to_string()));
        assert!(is_solvable(s(&["SIX", "SEVEN", "SEVEN"]), "TWENTY".to_string()));
        assert!(!is_solvable(s(&["LEET", "CODE"]), "POINT".to_string()));
        assert!(is_solvable(s(&["A", "B"]), "C".to_string()));
        assert!(!is_solvable(s(&["ACA"]), "JA".to_string()));
    }
}
